sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample11_MockingClientTypes.md

```diff
@@ -0,0 +1,292 @@
+
+# Mocking Client Types
+Event Hubs is built to be entirely mockable. This is an important feature of the library that allows developers to write tests that are completely focused on their own application logic, even if they are dependent on the Event Hubs types.
+
+The following snippets demonstrate how to mock each client type and relevant data types. 
+
+Another important focus of this document is guidance on abstracting . The Event Hubs library has an extensive test suite that verifies all types, clients, and methods work as specified. Each scenario also discusses how to leverage these guarantees provided by our library so that tests can just focus on what an application has layered on top.
+
+## Mocking `EventDataBatch`, `EventData`, and `EventHubProducerClient`
+
+When using batches to send to Event Hubs, many applications have specific needs for creating batches and preparing to send them. When testing, applications only need to focus their testing on the creation of batches, and the call to send. Once any of the send methods found with the `EventHubProducerClient` are called it can be assumed that the library will perform as expected.
+
+The following snippet demonstrates how to mock an `EventHubProducerClient` using Moq, and `EventDataBatch` and `EventData` instances using the `EventHubsModelFactory`. Mocked data batches can be used to test that an application is adding events to the batch properly. The custom `TryAdd` callback can be used to test all code paths for adding events to the batch. 
+
+The mocked `EventHubProducerClient` can be used to verify that `SendAsync` was called. The sample below is not a complete test. It can be customized to call an application-defined method that contains a call to `SendAsync`, and the arguments to `Verify` can be adjusted to match the expected outcome.
+
+```C# Snippet:EventHubs_Sample11_MockingEventDataBatch
+var mockProducer = new Mock<EventHubProducerClient>();
+
+var createBatchOptions = new CreateBatchOptions() { MaximumSizeInBytes = 516 };
+var batchSizeInBytes = 500;
+
+// Setting up a mock of the CreateBatchAsync method
+mockProducer.Setup(p => p.CreateBatchAsync(
+    It.IsAny<CancellationToken>()))
+    .ReturnsAsync(
+    EventHubsModelFactory.EventDataBatch(
+        46,
+        new List<EventData>(),
+        new CreateBatchOptions() { },
+        // The model factory allows the user to define a custom TryAdd callback, making
+        // it easy to test specific scenarios
+        eventData =>
+        {
+            return eventData.Body.Length > createBatchOptions.MaximumSizeInBytes - batchSizeInBytes;
+        }));
+
+// Mocking the SendAsync method so that it will always pass
+mockProducer.Setup(p => p.SendAsync(
+    It.IsAny<EventDataBatch>(),
```

Comment:
My thought with this scenario is that it would be a template for testing some application defined method that called `SendAsync` within it. 
Before simplifying the scenario, I called a sample method that contained a `SendAsync` call and then the test tested that it was called once. I'm assuming that people won't be calling `SendAsync` directly in their test method and instead calling some application defined method, but I can just add a call to send here for illustrative purposes.